modules/cactus-form/src/FormSpy.tsx

```diff
@@ -0,0 +1,60 @@
+import { FormState, formSubscriptionItems } from 'final-form'
+import PropTypes from 'prop-types'
+import React from 'react'
+import { FormSpyProps, FormSpyRenderProps, useForm, useFormState } from 'react-final-form'
+
+const DEFAULT_SUBSCRIPTION: Record<string, boolean> = {}
+formSubscriptionItems.forEach((sub) => {
+  DEFAULT_SUBSCRIPTION[sub] = true
+})
+
+type State = FormState<Record<string, any>, Partial<Record<string, any>>>
+type RestPropsWithChildren = Pick<FormSpyProps, 'onChange'> &
+  State & {
+    children?: React.ReactNode
+  }
+
+function FormSpy({ render, component, children, subscription, onChange, ...rest }: FormSpyProps) {
+  const form = useForm('FormSpy')
+  const sub = subscription || DEFAULT_SUBSCRIPTION
+  const [state, setState] = React.useState<State>(() => {
+    let initialState
+    form.subscribe((s) => (initialState = s), sub)()
+    return initialState as unknown as State
+  })
+  React.useEffect(() => {
+    const unsub = form.subscribe(setState, sub)
+
+    return () => unsub()
+  }, []) // eslint-disable-line react-hooks/exhaustive-deps
+
+  useFormState({ onChange, subscription: sub })
```

Comment:
The way you're setting the initial state is perfect, but this could be simplified a bit. In particular, using `useFormState` almost defeats the purpose of writing this component.
```suggestion
  React.useEffect(() => {
    return form.subscribe(onChange || setState, sub)
  }, [onChange]) // eslint-disable-line react-hooks/exhaustive-deps
```
Why do it this way compared to how they have it in React Final Form (with the ref for `onChange`)? It's best not to make assumptions about why the `onChange` function has changed. It might have changed simply because it's defined inline, in which case it would be better not to have it in the dependency array, but it might have changed for another reason as well, in which case we need to _make sure_ the upstream has the latest version of state by re-subscribing. You can put in the docs a recommendation to wrap `onChange` in `useCallback` if it's defined inline.

Reply:
Honestly I've never even come across a use case for `onChange` anyway, so it barely matters...

Reply:
Man, seeing your suggestion to just pass `onChange || setState` to `form.subscribe` makes me facepalm that I didn't see that connection before.

And you make a good point about putting `onChange` in the dependency array. It may be slightly less performant, but yeah it's probably best to avoid making assumptions. I'll make a note about wrapping it in `useCallback` in the docs.